File: 162. Find Peak Element.java
class Solution {
//     public int findPeakElement(int[] nums) {
//         if (nums.length == 1) return 0;
        
//         for (int i = 1; i < nums.length - 1; i++) {
//             if (nums[i] > nums[i-1] & nums[i] > nums[i+1]) {
//                 return i;
//             }
//         }
//         if (nums[0] > nums[1]) return 0;
//         if (nums[nums.length-1] > nums[nums.length-2]) return nums.length - 1;
        
//         return -1;
//     }
    
// method 2:
//     public int findPeakElement(int[] nums) {
//         return search(nums,0,nums.length - 1);
//     }
    
//     public int search(int[] nums, int l, int r) {
//         if (l == r) 
//             return l;
//         int mid = l + (r - l) / 2;
//         if (nums[mid] > nums[mid + 1])
//             return search(nums,l,mid);
//         return search(nums,mid+1,r);
//     }
    
    
// method 3:
    
    public int findPeakElement(int[] nums) {
        int l = 0, r = nums.length - 1;
        while (l < r) {
            int mid = l + (r - l)/2;
            if (nums[mid] > nums[mid + 1]) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        
        return l;
    }
}

File: amazon/python/162. Find Peak Element.py
class Solution(object):
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return -1
        if len(nums) == 1:
            return 0
        if len(nums) == 2:
            return 0 if nums[0] > nums[1] else 1
        
        if nums[0] > nums[1]:
            return 0
        
        if nums[len(nums) -1] > nums[len(nums) - 2]:
            return len(nums) - 1
        
        l = 0
        r = len(nums) - 1 # len(nums) >= 3
        while l <= r: # 4 cases 
            mid = l + (r-l)//2
            if nums[mid] > nums[mid - 1] and nums[mid] > nums[mid + 1]:
                return mid
            elif nums[mid] > nums[mid - 1] and nums[mid] < nums[mid + 1]:
                # ascending
                l = mid + 1
            elif nums[mid] < nums[mid - 1] and nums[mid] > nums[mid + 1]:
                # descending 
                r = mid - 1
            else: # at the valley
                l = l + 1
                # boundings 
        return -1
